File: kernel/runtime_modules/filesystem.ts
import { type Option, hasValue } from "ely:container";

/** Thrown when a userland filesystem path is given relative instead of
 * absolute. Every such path must start with `/` and is resolved against the
 * userland root.
 * `import.meta.directoryName`/`fileName` are how a module finds its own location
 * to build an absolute path from. */
export class RelativePathError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "RelativePathError";
  }
}

export const sep = "/";

function toPosixPath(path: string): string {
  return path.replace(/\\/g, sep);
}

/** Resolves `.`/`..` segments and collapses redundant slashes. The result is
 * always rooted at `/`; a `..` that would go above the root is dropped
 * rather than escaping it. */
export function normalize(path: string): string {
  const parts = toPosixPath(path).split(sep);
  const stack: string[] = [];

  for (const part of parts) {
    if (part === "" || part === ".") continue;
    if (part === "..") {
      if (stack.length > 0) stack.pop();
    } else {
      stack.push(part);
    }
  }

  return sep + stack.join(sep);
}

/** Joins path segments into one and normalizes the result. */
export function join(...paths: string[]): string {
  return normalize(paths.join(sep));
}

/** The final segment of a path, with `ext` stripped from the end if
 * present. Trailing slashes are ignored. */
export function extractBaseName(path: string, ext: string = ""): string {
  path = toPosixPath(path);
  let end = path.length;
  while (end > 1 && path[end - 1] === sep) end--;
  path = path.slice(0, end);

  const base = path.substring(path.lastIndexOf(sep) + 1);
  if (ext && base.endsWith(ext)) {
    return base.slice(0, -ext.length);
  }
  return base;
}

/** The path up to, but not including, the final segment. Trailing slashes
 * are ignored, and the root's own parent is itself. */
export function extractDirectoryName(path: string): string {
  path = toPosixPath(path);
  let end = path.length;
  while (end > 1 && path[end - 1] === sep) end--;
  path = path.slice(0, end);

  if (path === sep) return sep;

  const idx = path.lastIndexOf(sep);
  if (idx === -1) return ".";
  if (idx === 0) return sep;
  return path.slice(0, idx);
}

/** The final segment's extension, including the leading `.`, or `""` if it
 * has none. A leading dot, as in `.bashrc`, is not itself an extension. */
export function extractExtension(path: string): string {
  const base = extractBaseName(path);
  const dotIndex = base.lastIndexOf(".");
  return dotIndex > 0 ? base.substring(dotIndex) : "";
}

/** Builds an absolute path by resolving segments right-to-left, stopping at
 * the first one that's already absolute. Equivalent to `join` when none of
 * the segments are absolute. */
export function resolve(...paths: string[]): string {
  let resolvedPath = "";
  let absolute = false;

  for (let i = paths.length - 1; i >= 0 && !absolute; i--) {
    const path = toPosixPath(paths[i] ?? "");
    if (!path) continue;

    resolvedPath = `${path}/${resolvedPath}`;
    absolute = path.startsWith(sep);
  }

  return normalize(resolvedPath);
}

/** Replaces characters outside `[a-zA-Z0-9_.-]` with `_` and truncates to
 * `maxLength`. `.` and `..` are replaced outright, since every one of their
 * characters is otherwise allowed and they would pass through unchanged and
 * remain usable for directory traversal. */
export function sanitizeName(
  filename: string,
  maxLength: number = 255,
): string {
  if (filename === "." || filename === "..") {
    return "_".repeat(filename.length);
  }

  const sanitized = filename.replace(/[^a-zA-Z0-9_\-.]/g, "_");
  return sanitized.slice(0, maxLength);
}

declare function __fs_read_file(path: string, offset: number, length: number): Uint8Array;
declare function __fs_write_file(
  path: string,
  data: Uint8Array,
  offset: number,
  truncate: boolean,
): void;
declare function __fs_read_text_file(path: string): string;
declare function __fs_write_text_file(path: string, text: string): void;
declare function __fs_remove(path: string): void;
declare function __fs_create_directory(path: string): void;
declare function __fs_list_directory(
  path: string,
): ({ kind: "File"; size: number; path: string } | { kind: "Directory"; path: string })[];
declare function __fs_stat(
  path: string,
): { kind: "File"; size: number; path: string } | { kind: "Directory"; path: string };

// The classes below are a shared vocabulary reused across every function —
// mirroring `std::io::ErrorKind` being one shared set of variants for every
// Rust filesystem call, rather than a bespoke error type per call site. Each
// function's exported error type (`ReadFileError`, `WriteFileError`, ...) is
// just the union of whichever of these it can actually throw.

/** A path, or a component of it, doesn't exist. */
export class NotFoundError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "NotFoundError";
  }
}

/** A file operation (read, write, or removing a file) found a directory
 * instead. */
export class IsADirectoryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "IsADirectoryError";
  }
}

/** A directory operation (`createDirectory`, `listDirectory`) is blocked by
 * an existing non-directory somewhere along the path. */
export class NotADirectoryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "NotADirectoryError";
  }
}

/** `readTextFile`'s bytes aren't valid UTF-8. */
export class TextDecodeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "TextDecodeError";
  }
}

/** Anything else the underlying operation failed with — permission denied,
 * disk full, and so on. `.message` carries the full underlying OS error
 * text plus the operation and path involved, so it stays useful for
 * debugging even though it isn't one of the specifically-typed cases
 * above. */
export class UnknownError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UnknownError";
  }
}

/** Parses the `"<TAG>: "` prefix `kernel/filesystem.rs` attaches to every
 * native error message and throws the matching shared class above, with
 * everything after the prefix — operation, path, and full OS error text —
 * intact as `.message`. Falls back to `UnknownError` for an
 * unrecognized/missing tag. */
function throwTagged(err: unknown): never {
  const raw = err instanceof Error ? err.message : String(err);
  const separator = raw.indexOf(": ");
  const tag = separator === -1 ? "" : raw.slice(0, separator);
  const message = separator === -1 ? raw : raw.slice(separator + 2);

  switch (tag) {
    case "NOT_FOUND":
      throw new NotFoundError(message);
    case "IS_A_DIRECTORY":
      throw new IsADirectoryError(message);
    case "NOT_A_DIRECTORY":
      throw new NotADirectoryError(message);
    case "INVALID_UTF8":
      throw new TextDecodeError(message);
    default:
      throw new UnknownError(message);
  }
}

/** Every case `readFile` can actually throw natively (beyond
 * `RelativePathError`, thrown before any native call). */
export type ReadFileError = NotFoundError | IsADirectoryError | UnknownError;

/** Every case `readTextFile` can actually throw natively — like
 * `ReadFileError`, plus `TextDecodeError` for non-UTF-8 bytes. */
export type ReadTextFileError =
  | NotFoundError
  | IsADirectoryError
  | TextDecodeError
  | UnknownError;

/** Every case `writeFile` can actually throw natively. */
export type WriteFileError = NotFoundError | IsADirectoryError | UnknownError;

/** Every case `writeTextFile` can actually throw natively. */
export type WriteTextFileError = NotFoundError | IsADirectoryError | UnknownError;

/** Every case `remove` can actually throw natively. */
export type DeleteError = NotFoundError | UnknownError;

/** Every case `createDirectory` can actually throw natively. */
export type CreateDirectoryError = NotADirectoryError | UnknownError;

/** Every case `listDirectory` can actually throw natively. */
export type ListDirectoryError = NotFoundError | NotADirectoryError | UnknownError;

/** Every case `stat` can actually throw natively. */
export type StatError = NotFoundError | UnknownError;

/** A byte range within a file, used by `readFile`/`writeFile` to operate on
 * part of a file instead of the whole thing. Both fields are optional:
 * `offset` defaults to `0`, `length` defaults to "the rest of the file" for
 * `readFile` and "all of `data`" for `writeFile`. */
export interface ByteRange {
  offset?: number;
  length?: number;
}

/** The result of `stat`, and of each entry `listDirectory` returns.
 * `path` is the entry's own absolute, virtual path — for `listDirectory`,
 * that's the listed directory joined with the entry's name, not just the
 * name by itself. */
export type EntryStat =
  | { readonly kind: "File"; readonly size: number; readonly path: string }
  | { readonly kind: "Directory"; readonly path: string };

function assertAbsolute(path: string): void {
  if (!path.startsWith("/")) {
    throw new RelativePathError(`${path} is not an absolute path`);
  }
}

/** Reads the whole file at `path`, or — if `range` is given — only the
 * bytes from `range.offset` (default `0`) spanning at most `range.length`
 * bytes (default: to the end of the file). `path` must be absolute.
 * @throws {RelativePathError} if `path` isn't absolute.
 * @throws {NotFoundError} if `path` doesn't exist.
 * @throws {IsADirectoryError} if `path` names a directory.
 * @throws {UnknownError} on any other failure. */
export function readFile(path: string, range?: Option<ByteRange>): Uint8Array {
  assertAbsolute(path);
  const offset = hasValue(range) ? (range.offset ?? 0) : 0;
  const length = hasValue(range) && range.length !== undefined ? range.length : -1;

  try {
    return __fs_read_file(path, offset, length);
  } catch (err) {
    throwTagged(err);
  }
}

/** Writes `data` to `path`. With no `range`, the file is truncated and
 * replaced entirely. With `range`, the file is patched in place: writing
 * starts at `range.offset` (default `0`), and bytes outside the written
 * span are left untouched. `path` must be absolute.
 * @throws {RelativePathError} if `path` isn't absolute.
 * @throws {NotFoundError} if `path`'s parent directory doesn't exist.
 * @throws {IsADirectoryError} if `path` names a directory.
 * @throws {UnknownError} on any other failure. */
export function writeFile(path: string, data: Uint8Array, range?: Option<ByteRange>): void {
  assertAbsolute(path);
  const offset = hasValue(range) ? (range.offset ?? 0) : 0;
  const length = hasValue(range) && range.length !== undefined ? range.length : data.length;
  const truncate = !hasValue(range);
  const slice = length === data.length ? data : data.slice(0, length);

  try {
    __fs_write_file(path, slice, offset, truncate);
  } catch (err) {
    throwTagged(err);
  }
}

// Userland has no `TextEncoder`/`TextDecoder` yet (planned as a polyfill on
// a separate branch), so these two do their own UTF-8 conversion natively
// rather than layering on top of `readFile`/`writeFile`. Once that polyfill
// lands, these should likely become thin wrappers around them instead.

/** Reads the file at `path` as UTF-8 text. `path` must be absolute.
 * @throws {RelativePathError} if `path` isn't absolute.
 * @throws {NotFoundError} if `path` doesn't exist.
 * @throws {IsADirectoryError} if `path` names a directory.
 * @throws {TextDecodeError} if `path`'s bytes aren't valid UTF-8.
 * @throws {UnknownError} on any other failure. */
export function readTextFile(path: string): string {
  assertAbsolute(path);
  try {
    return __fs_read_text_file(path);
  } catch (err) {
    throwTagged(err);
  }
}

/** Writes `text` to `path`, truncating and replacing it entirely. `path`
 * must be absolute.
 * @throws {RelativePathError} if `path` isn't absolute.
 * @throws {NotFoundError} if `path`'s parent directory doesn't exist.
 * @throws {IsADirectoryError} if `path` names a directory.
 * @throws {UnknownError} on any other failure. */
export function writeTextFile(path: string, text: string): void {
  assertAbsolute(path);
  try {
    __fs_write_text_file(path, text);
  } catch (err) {
    throwTagged(err);
  }
}

/** Removes the file or directory at `path`. A directory is removed
 * recursively, along with everything inside it. `path` must be absolute.
 * @throws {RelativePathError} if `path` isn't absolute.
 * @throws {NotFoundError} if `path` doesn't exist.
 * @throws {UnknownError} on any other failure. */
export function remove(path: string): void {
  assertAbsolute(path);
  try {
    __fs_remove(path);
  } catch (err) {
    throwTagged(err);
  }
}

/** Creates the directory at `path`, along with any missing parent
 * directories. `path` must be absolute.
 * @throws {RelativePathError} if `path` isn't absolute.
 * @throws {NotADirectoryError} if a non-directory already exists somewhere
 * along `path`.
 * @throws {UnknownError} on any other failure. */
export function createDirectory(path: string): void {
  assertAbsolute(path);
  try {
    __fs_create_directory(path);
  } catch (err) {
    throwTagged(err);
  }
}

/** Lists the entries directly inside the directory at `path`, each as an
 * `EntryStat` whose `path` is that entry's own absolute path (the listed
 * directory joined with its name). `path` must be absolute.
 * @throws {RelativePathError} if `path` isn't absolute.
 * @throws {NotFoundError} if `path` doesn't exist.
 * @throws {NotADirectoryError} if `path` names a file.
 * @throws {UnknownError} on any other failure. */
export function listDirectory(path: string): EntryStat[] {
  assertAbsolute(path);
  try {
    return __fs_list_directory(path);
  } catch (err) {
    throwTagged(err);
  }
}

/** Reports whether `path` is a file or a directory, its size in bytes if
 * it's a file, and its own absolute path. `path` must be absolute.
 * @throws {RelativePathError} if `path` isn't absolute.
 * @throws {NotFoundError} if `path` doesn't exist.
 * @throws {UnknownError} on any other failure. */
export function stat(path: string): EntryStat {
  assertAbsolute(path);
  try {
    return __fs_stat(path);
  } catch (err) {
    throwTagged(err);
  }
}

File: kernel/filesystem.rs
//! File and directory I/O backing `ely:filesystem`'s native bindings.
//! Unlike `kernel/image.rs`, no per-VM resource table is needed here — every
//! operation (read, write, stat, list, create, remove) is one-shot, so
//! `bootstrap_filesystem_bindings` only ever captures a cloned
//! `userland_root`.
//!
//! Symlinks are made fully invisible rather than transparently followed:
//! [`resolve_userland_path`] and [`resolve_userland_path_for_write`] walk a
//! requested path component by component from the userland root, checking
//! `std::fs::symlink_metadata` at every step, and treat any symlink
//! component as though the path didn't exist at all. Both are `pub`: they
//! are the one place userland-path sandboxing is implemented, and every
//! other native surface that resolves a virtual path — `ely:image`'s
//! `loadImage`, the process manager's program loading — goes through them
//! rather than rolling its own `canonicalize`-based check.
//!
//! Every thrown message is tagged with a machine-readable prefix
//! (`"NOT_FOUND: ..."`, `"IS_A_DIRECTORY: ..."`, ...) that `ely:filesystem`
//! parses to pick which of its specific, shared error classes
//! (`NotFoundError`, `IsADirectoryError`, ...) to construct — see
//! [`throw_resolution`]/[`throw_io`]. The tag is a prefix added on top of a
//! normal, fully-detailed message (operation, virtual path, and the
//! underlying OS error text), never a replacement for it.

use std::io::ErrorKind;
use std::path::{Component, Path, PathBuf};

use crate::bindings::bind;
use rquickjs::{Ctx, Object, Result, TypedArray};

/// Why [`resolve_userland_path`]/[`resolve_userland_path_for_write`] failed
/// to resolve a requested path, carrying enough to build both a tag and a
/// rich message — see [`throw_resolution`].
pub enum PathResolutionError {
    /// `requested` didn't start with `/`. In practice unreachable from
    /// `ely:filesystem`'s exported functions, which all call
    /// `assertAbsolute` before ever reaching a native binding — kept only
    /// as defense in depth for a binding called some other way.
    NotAbsolute(String),
    /// A component of `requested` doesn't exist, or is a symlink (symlinks
    /// are invisible to the VM, so this is indistinguishable from not
    /// existing).
    NotFound(String),
    /// A component of `requested` isn't a normal path segment (e.g. a
    /// Windows-style drive prefix) — not reachable from a POSIX-style
    /// virtual path, kept only for exhaustiveness.
    Invalid(String),
}

impl PathResolutionError {
    fn tag(&self) -> &'static str {
        match self {
            PathResolutionError::NotAbsolute(_) => "NOT_FOUND",
            PathResolutionError::NotFound(_) => "NOT_FOUND",
            PathResolutionError::Invalid(_) => "IO_ERROR",
        }
    }

    fn message(&self) -> &str {
        match self {
            PathResolutionError::NotAbsolute(message)
            | PathResolutionError::NotFound(message)
            | PathResolutionError::Invalid(message) => message,
        }
    }
}

impl std::fmt::Display for PathResolutionError {
    /// The rich message without the machine-readable tag — for callers
    /// outside `ely:filesystem` (e.g. `ely:image`, program loading) that
    /// surface a plain error rather than parsing tags.
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.message())
    }
}

/// Throws `err` as a tagged JS exception — see the module doc comment.
fn throw_resolution(ctx: &Ctx<'_>, err: PathResolutionError) -> rquickjs::Error {
    rquickjs::Exception::throw_message(ctx, &format!("{}: {}", err.tag(), err.message()))
}

/// Tags a `std::io::Error` from `operation` on `path` and throws it as a
/// tagged JS exception, keeping the full `path`/`operation`/OS-error detail
/// in the message — see the module doc comment.
fn throw_io(ctx: &Ctx<'_>, operation: &str, path: &str, err: std::io::Error) -> rquickjs::Error {
    let tag = match err.kind() {
        ErrorKind::NotFound => "NOT_FOUND",
        ErrorKind::IsADirectory => "IS_A_DIRECTORY",
        ErrorKind::NotADirectory => "NOT_A_DIRECTORY",
        _ => "IO_ERROR",
    };
    rquickjs::Exception::throw_message(ctx, &format!("{tag}: {operation} {path}: {err}"))
}

/// Like [`throw_io`], but for `std::fs::read_to_string` specifically: its
/// `ErrorKind::InvalidData` means the file's bytes aren't valid UTF-8,
/// which every other caller's `ErrorKind` set doesn't need to distinguish.
fn throw_read_to_string_io(ctx: &Ctx<'_>, path: &str, err: std::io::Error) -> rquickjs::Error {
    if err.kind() == ErrorKind::InvalidData {
        return rquickjs::Exception::throw_message(
            ctx,
            &format!("INVALID_UTF8: read {path}: {err}"),
        );
    }
    throw_io(ctx, "read", path, err)
}

/// Resolves `requested` — an absolute, virtual, userland-rooted path — to
/// its real path, requiring every component (including the leaf) to already
/// exist and none of them to be a symlink. `canonical_userland_root` must
/// already be canonicalized (as [`crate::runtime::ElysiumRuntime::new`]
/// does once for the whole VM); by construction the walk never produces a
/// path outside it, since `..` is clamped at the root exactly like
/// `ely:filesystem`'s own `normalize()`.
pub fn resolve_userland_path(
    canonical_userland_root: &Path,
    requested: &str,
) -> std::result::Result<PathBuf, PathResolutionError> {
    walk_userland_path(canonical_userland_root, requested, false)
}

/// Like [`resolve_userland_path`], but tolerates `requested`'s leaf (and any
/// number of trailing components) not existing yet — needed by
/// `createDirectory` (which may need to create several missing levels) and
/// by a `writeFile` targeting a not-yet-existing file.
pub fn resolve_userland_path_for_write(
    canonical_userland_root: &Path,
    requested: &str,
) -> std::result::Result<PathBuf, PathResolutionError> {
    walk_userland_path(canonical_userland_root, requested, true)
}

/// Shared walk behind [`resolve_userland_path`] and
/// [`resolve_userland_path_for_write`]. Advances component by component from
/// `canonical_userland_root`, checking `symlink_metadata` at every step and
/// treating any symlink component as though it didn't exist, and clamps
/// `..` at the root so the result can never escape it.
///
/// When a component doesn't exist: if `allow_missing_tail` is false the
/// whole path is rejected as not found; if it is true, that component and
/// everything after it is appended lexically (still clamping `..`) without
/// touching the filesystem again — a component that doesn't exist yet can't
/// itself be a symlink, so no further check is needed past that point.
fn walk_userland_path(
    canonical_userland_root: &Path,
    requested: &str,
    allow_missing_tail: bool,
) -> std::result::Result<PathBuf, PathResolutionError> {
    let Some(relative) = requested.strip_prefix('/') else {
        return Err(PathResolutionError::NotAbsolute(format!(
            "{requested} is not an absolute path"
        )));
    };

    let not_found =
        || PathResolutionError::NotFound(format!("{requested}: no such file or directory"));
    let invalid = || PathResolutionError::Invalid(format!("{requested} is not a valid path"));

    let mut current = canonical_userland_root.to_path_buf();
    let mut components = Path::new(relative).components();

    while let Some(component) = components.next() {
        match component {
            Component::Normal(part) => {
                let candidate = current.join(part);
                match std::fs::symlink_metadata(&candidate) {
                    Ok(metadata) => {
                        if metadata.file_type().is_symlink() {
                            return Err(not_found());
                        }
                        current = candidate;
                    }
                    Err(_) if allow_missing_tail => {
                        current.push(part);
                        for remaining in components {
                            match remaining {
                                Component::Normal(part) => current.push(part),
                                Component::ParentDir => {
                                    if current != canonical_userland_root {
                                        current.pop();
                                    }
                                }
                                Component::CurDir | Component::RootDir => {}
                                _ => return Err(invalid()),
                            }
                        }
                        return Ok(current);
                    }
                    Err(_) => return Err(not_found()),
                }
            }
            Component::ParentDir => {
                if current != canonical_userland_root {
                    current.pop();
                }
            }
            Component::CurDir | Component::RootDir => {}
            _ => return Err(invalid()),
        }
    }

    Ok(current)
}

/// Expresses `resolved` — a real, absolute path known to be inside
/// `canonical_userland_root` — as a virtual, userland-rooted path (e.g.
/// `/programs/init/index.ts`), the form `EntryStat.path` is reported in.
fn virtual_path(canonical_userland_root: &Path, resolved: &Path) -> String {
    let relative = resolved
        .strip_prefix(canonical_userland_root)
        .unwrap_or(resolved);
    format!("/{}", relative.to_string_lossy())
}

/// Reads `resolved`'s bytes, optionally restricted to the range starting at
/// `offset` and spanning at most `length` bytes (`length < 0` means "to the
/// end of the file"). Used by `__fs_read_file`.
fn read_range(resolved: &Path, offset: i64, length: i64) -> std::io::Result<Vec<u8>> {
    let bytes = std::fs::read(resolved)?;
    let offset = offset.max(0) as usize;
    if offset >= bytes.len() {
        return Ok(Vec::new());
    }
    let end = if length < 0 {
        bytes.len()
    } else {
        bytes.len().min(offset.saturating_add(length as usize))
    };
    Ok(bytes[offset..end].to_vec())
}

/// Writes `data` into `resolved` at `offset` (bytes before it are left
/// untouched). When `truncate` is set, the file is truncated to exactly
/// `data`'s length first — the "no range given, replace the whole file"
/// call `writeFile` makes when its caller didn't pass a `ByteRange`. Used by
/// `__fs_write_file`.
fn write_range(resolved: &Path, data: &[u8], offset: i64, truncate: bool) -> std::io::Result<()> {
    use std::io::{Seek, SeekFrom, Write};

    let mut file = std::fs::OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(truncate)
        .open(resolved)?;
    file.seek(SeekFrom::Start(offset.max(0) as u64))?;
    file.write_all(data)
}

/// Resolves `path` for reading, throwing the tagged JS exception on failure —
/// the opening move of every binding below that requires its target to exist.
fn resolve_for_read(ctx: &Ctx<'_>, root: &Path, path: &str) -> Result<PathBuf> {
    resolve_userland_path(root, path).map_err(|err| throw_resolution(ctx, err))
}

/// Like [`resolve_for_read`], but for the bindings that may be naming
/// something not created yet.
fn resolve_for_write(ctx: &Ctx<'_>, root: &Path, path: &str) -> Result<PathBuf> {
    resolve_userland_path_for_write(root, path).map_err(|err| throw_resolution(ctx, err))
}

/// Binds the *hidden* globals `ely:filesystem`'s embedded module wraps
/// (`__fs_read_file`, `__fs_write_file`, ...) — never called by a program
/// directly, only through `ely:filesystem`'s exported functions.
/// `userland_root` is already canonicalized by the caller
/// ([`crate::runtime::ElysiumRuntime::new`]).
pub fn bootstrap_filesystem_bindings<'js>(ctx: &Ctx<'js>, userland_root: PathBuf) -> Result<()> {
    {
        let userland_root = userland_root.clone();
        bind(
            ctx,
            "__fs_read_file",
            move |ctx: Ctx<'js>,
                  path: String,
                  offset: i64,
                  length: i64|
                  -> Result<TypedArray<'js, u8>> {
                let resolved = resolve_for_read(&ctx, &userland_root, &path)?;
                let bytes = read_range(&resolved, offset, length)
                    .map_err(|err| throw_io(&ctx, "read", &path, err))?;
                TypedArray::new_copy(ctx.clone(), bytes)
            },
        )?;
    }

    {
        let userland_root = userland_root.clone();
        bind(
            ctx,
            "__fs_write_file",
            move |ctx: Ctx<'_>,
                  path: String,
                  data: TypedArray<'_, u8>,
                  offset: i64,
                  truncate: bool|
                  -> Result<()> {
                let resolved = resolve_for_write(&ctx, &userland_root, &path)?;
                let bytes = data.as_bytes().ok_or_else(|| {
                    rquickjs::Exception::throw_message(
                        &ctx,
                        &format!("IO_ERROR: write {path}: data is not backed by an ArrayBuffer"),
                    )
                })?;
                write_range(&resolved, bytes, offset, truncate)
                    .map_err(|err| throw_io(&ctx, "write", &path, err))
            },
        )?;
    }

    {
        let userland_root = userland_root.clone();
        bind(
            ctx,
            "__fs_read_text_file",
            move |ctx: Ctx<'_>, path: String| -> Result<String> {
                let resolved = resolve_for_read(&ctx, &userland_root, &path)?;
                std::fs::read_to_string(&resolved)
                    .map_err(|err| throw_read_to_string_io(&ctx, &path, err))
            },
        )?;
    }

    {
        let userland_root = userland_root.clone();
        bind(
            ctx,
            "__fs_write_text_file",
            move |ctx: Ctx<'_>, path: String, text: String| -> Result<()> {
                let resolved = resolve_for_write(&ctx, &userland_root, &path)?;
                std::fs::write(&resolved, text).map_err(|err| throw_io(&ctx, "write", &path, err))
            },
        )?;
    }

    {
        let userland_root = userland_root.clone();
        bind(
            ctx,
            "__fs_remove",
            move |ctx: Ctx<'_>, path: String| -> Result<()> {
                let resolved = resolve_for_read(&ctx, &userland_root, &path)?;
                let metadata = std::fs::metadata(&resolved)
                    .map_err(|err| throw_io(&ctx, "remove", &path, err))?;
                let result = if metadata.is_dir() {
                    std::fs::remove_dir_all(&resolved)
                } else {
                    std::fs::remove_file(&resolved)
                };
                result.map_err(|err| throw_io(&ctx, "remove", &path, err))
            },
        )?;
    }

    {
        let userland_root = userland_root.clone();
        bind(
            ctx,
            "__fs_create_directory",
            move |ctx: Ctx<'_>, path: String| -> Result<()> {
                let resolved = resolve_for_write(&ctx, &userland_root, &path)?;
                std::fs::create_dir_all(&resolved)
                    .map_err(|err| throw_io(&ctx, "create directory", &path, err))
            },
        )?;
    }

    {
        let userland_root = userland_root.clone();
        bind(
            ctx,
            "__fs_list_directory",
            move |ctx: Ctx<'js>, path: String| -> Result<Vec<Object<'js>>> {
                let resolved = resolve_for_read(&ctx, &userland_root, &path)?;
                let entries = std::fs::read_dir(&resolved)
                    .map_err(|err| throw_io(&ctx, "list directory", &path, err))?;

                let mut result = Vec::new();
                for entry in entries {
                    let entry =
                        entry.map_err(|err| throw_io(&ctx, "list directory", &path, err))?;
                    let file_type = entry
                        .file_type()
                        .map_err(|err| throw_io(&ctx, "list directory", &path, err))?;
                    // Only files and directories are userland-visible —
                    // symlinks are invisible (see the module doc
                    // comment) and anything else (sockets, devices,
                    // ...) is out of scope for a files-and-directories
                    // filesystem API.
                    if !file_type.is_file() && !file_type.is_dir() {
                        continue;
                    }

                    let entry_path = virtual_path(&userland_root, &entry.path());
                    let object = Object::new(ctx.clone())?;
                    if file_type.is_dir() {
                        object.set("kind", "Directory")?;
                    } else {
                        let metadata = entry
                            .metadata()
                            .map_err(|err| throw_io(&ctx, "list directory", &path, err))?;
                        object.set("kind", "File")?;
                        object.set("size", metadata.len() as f64)?;
                    }
                    object.set("path", entry_path)?;
                    result.push(object);
                }
                Ok(result)
            },
        )?;
    }

    bind(
        ctx,
        "__fs_stat",
        move |ctx: Ctx<'js>, path: String| -> Result<Object<'js>> {
            let resolved = resolve_for_read(&ctx, &userland_root, &path)?;
            let metadata =
                std::fs::metadata(&resolved).map_err(|err| throw_io(&ctx, "stat", &path, err))?;

            let object = Object::new(ctx.clone())?;
            if metadata.is_dir() {
                object.set("kind", "Directory")?;
            } else if metadata.is_file() {
                object.set("kind", "File")?;
                object.set("size", metadata.len() as f64)?;
            } else {
                // Sockets, devices, etc. — outside the files-and-directories
                // scope `ely:filesystem` covers.
                return Err(rquickjs::Exception::throw_message(
                    &ctx,
                    &format!("IO_ERROR: stat {path}: not a file or directory"),
                ));
            }
            object.set("path", virtual_path(&userland_root, &resolved))?;
            Ok(object)
        },
    )?;

    Ok(())
}
